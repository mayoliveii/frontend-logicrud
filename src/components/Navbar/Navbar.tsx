'use client';
import {
  Box,
  Flex,
  Avatar,
  HStack,
  Image,
  IconButton,
  Button,
  Menu,
  MenuButton,
  MenuList,
  MenuItem,
  MenuDivider,
  useDisclosure,
  useColorModeValue,
  Stack,
  Text,
  Link as ChakraLink,
} from '@chakra-ui/react';
import { HamburgerIcon, CloseIcon, AddIcon } from '@chakra-ui/icons';
import { FaLinkedin } from 'react-icons/fa';

interface NavLinkProps {
  children: React.ReactNode;
  href: string;
}

export const NavLink = ({ children, href, noHover }: NavLinkProps & { noHover?: boolean }) => {
  const bgColor = useColorModeValue('gray.200', 'gray.700');

  return (
    <Box
      as="a"
      px={2}
      py={1}
      rounded={'md'}
      _hover={noHover ? {} : {
        textDecoration: 'none',
        bg: bgColor,
      }}
      href={href}>
      {children}
    </Box>
  );
};

const Links = [
  { name: 'Veículos', path: '/vehicle-list' },
  { name: 'Sobre', path: '/about' },
  { name: 'Time', path: '/team' },
];

export const MenuComponent = () => (
  <Menu>
    <MenuButton
      as={Button}
      rounded={'full'}
      variant={'link'}
      cursor={'pointer'}
      minW={0}>
      <Avatar
        size={'sm'}
        src={'/images/avatar-3.png'}
        style={{ maxWidth: '40px', maxHeight: '40px', borderRadius: '50%' }}
      />
    </MenuButton>
    <MenuList>
      <Box p={2}>
        <Text>Mayara Oliveira</Text>
        <Text fontSize="sm" color="gray.500">Desenvolvedora de Software</Text>
      </Box>
      <MenuDivider />
      <MenuItem>
        <ChakraLink href="https://www.linkedin.com/in/mayoliveii/" isExternal>
          <HStack spacing={2}>
            <FaLinkedin />
            <Text>LinkedIn</Text>
          </HStack>
        </ChakraLink>
      </MenuItem>
    </MenuList>
  </Menu>
);

export const IconButtonComponent = ({ isOpen, onOpen, onClose }: { isOpen: boolean, onOpen: () => void, onClose: () => void }) => (
  <IconButton
    size={'md'}
    icon={isOpen ? <CloseIcon /> : <HamburgerIcon />}
    aria-label={'Open Menu'}
    display={{ md: 'none' }}
    onClick={isOpen ? onClose : onOpen}
  />
);

export default function Navbar() {
  const { isOpen, onOpen, onClose } = useDisclosure();

  return (
    <>
      <Box px={4} m={1}>
        <Flex h={16} alignItems={'center'} justifyContent={'space-between'}>
          <IconButtonComponent isOpen={isOpen} onOpen={onOpen} onClose={onClose} />
          <HStack spacing={8} alignItems={'center'}>
            <NavLink noHover href="/">
              <Image
                src="/images/logo.png"
                alt="Logo"
                width="150px"
                height="80px"
              />
            </NavLink>
            <HStack as={'nav'} spacing={4} display={{ base: 'none', md: 'flex' }}>
              {Links.map((link) => (
                <NavLink key={link.name} href={link.path}>
                  {link.name}
                </NavLink>
              ))}
            </HStack>
          </HStack>
          <Flex alignItems={'center'}>
            <ChakraLink href="https://verticalti.com.br/#solucoes" isExternal>
              <Button
                variant={'solid'}
                bg={'#FF7A50'}
                _hover={{ bg: '#FF5823' }}
                color={'white'}
                size={'sm'}
                mr={4}
                leftIcon={<AddIcon />}>
                Soluções
              </Button>
            </ChakraLink>
            <MenuComponent />
          </Flex>
        </Flex>

        {isOpen ? (
          <Box pb={4} display={{ md: 'none' }}>
            <Stack as={'nav'} spacing={4}>
              {Links.map((link) => (
                <NavLink key={link.name} href={link.path}>
                  {link.name}
                </NavLink>
              ))}
            </Stack>
          </Box>
        ) : null}
      </Box>
    </>
  );
}