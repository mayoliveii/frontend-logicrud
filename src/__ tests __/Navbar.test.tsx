import { render, fireEvent } from '@testing-library/react';
import { ChakraProvider } from '@chakra-ui/react';
import Navbar, { IconButtonComponent, MenuComponent, NavLink } from '../components/Navbar/Navbar';
import '@testing-library/jest-dom';

describe('Navbar', () => {
  it('renders snapshot correctly', () => {
    const { asFragment } = render(
      <ChakraProvider>
        <Navbar />
      </ChakraProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders NavLink correctly', () => {
    const { getByText } = render(
      <ChakraProvider>
        <NavLink href="/test">Test Link</NavLink>
      </ChakraProvider>
    );
    expect(getByText('Test Link')).toBeInTheDocument();
  });

  it('renders MenuComponent correctly', () => {
    const { getByText } = render(
      <ChakraProvider>
        <MenuComponent />
      </ChakraProvider>
    );
    expect(getByText('Mayara Oliveira')).toBeInTheDocument();
    expect(getByText('Desenvolvedora de Software')).toBeInTheDocument();
  });

  it('renders IconButtonComponent correctly and responds to click events', () => {
    const mockOnOpen = jest.fn();
    const mockOnClose = jest.fn();
    const { getByLabelText } = render(
      <ChakraProvider>
        <IconButtonComponent isOpen={false} onOpen={mockOnOpen} onClose={mockOnClose} />
      </ChakraProvider>
    );
    fireEvent.click(getByLabelText('Open Menu'));
    expect(mockOnOpen).toHaveBeenCalled();
  });

  it('renders Navbar correctly', () => {
    const { getByText } = render(
      <ChakraProvider>
        <Navbar />
      </ChakraProvider>
    );
    expect(getByText('Ve√≠culos')).toBeInTheDocument();
    expect(getByText('Sobre')).toBeInTheDocument();
    expect(getByText('Time')).toBeInTheDocument();
  });
});