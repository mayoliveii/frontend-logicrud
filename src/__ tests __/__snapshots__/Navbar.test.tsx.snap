// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders snapshot correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1lrs3p7"
  >
    <div
      class="css-19pkskf"
    >
      <button
        aria-label="Open Menu"
        class="chakra-button css-n1aaeq"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="chakra-stack css-pa7cn"
      >
        <a
          class="css-3sczcp"
          href="/"
        >
          <img
            alt="Logo"
            class="chakra-image css-vjb3xi"
            src="https://mayoliveii.s3.us-east-2.amazonaws.com/images/logicrud-logo.jpg"
          />
        </a>
        <nav
          class="chakra-stack css-1y6oau7"
        >
          <a
            class="css-3yj86i"
            href="/vehicle-list"
          >
            Veículos
          </a>
          <a
            class="css-3yj86i"
            href="/about"
          >
            Sobre
          </a>
          <a
            class="css-3yj86i"
            href="/team"
          >
            Time
          </a>
        </nav>
      </div>
      <div
        class="css-70qvj9"
      >
        <a
          class="chakra-link css-spn4bz"
          href="https://verticalti.com.br/#solucoes"
          rel="noopener"
          target="_blank"
        >
          <button
            class="chakra-button css-4s8921"
            type="button"
          >
            <span
              class="chakra-button__icon css-1wh2kri"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M0,12a1.5,1.5,0,0,0,1.5,1.5h8.75a.25.25,0,0,1,.25.25V22.5a1.5,1.5,0,0,0,3,0V13.75a.25.25,0,0,1,.25-.25H22.5a1.5,1.5,0,0,0,0-3H13.75a.25.25,0,0,1-.25-.25V1.5a1.5,1.5,0,0,0-3,0v8.75a.25.25,0,0,1-.25.25H1.5A1.5,1.5,0,0,0,0,12Z"
                  fill="currentColor"
                />
              </svg>
            </span>
            Soluções
          </button>
        </a>
        <button
          aria-controls="menu-list-:r2:"
          aria-expanded="false"
          aria-haspopup="menu"
          class="chakra-button chakra-menu__menu-button css-7pfzb9"
          id="menu-button-:r2:"
          type="button"
        >
          <span
            class="css-xl71ch"
          >
            <span
              class="chakra-avatar css-j0jwkr"
              style="max-width: 40px; max-height: 40px; border-radius: 50%;"
            >
              <svg
                aria-label=" avatar"
                class="chakra-avatar__svg css-16ite8i"
                role="img"
                viewBox="0 0 128 128"
              >
                <path
                  d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
                  fill="currentColor"
                />
                <path
                  d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
                  fill="currentColor"
                />
              </svg>
            </span>
          </span>
        </button>
        <div
          class="css-r6z5ec"
          style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
        >
          <div
            aria-orientation="vertical"
            class="chakra-menu__menu-list css-1kfu8nn"
            id="menu-list-:r2:"
            role="menu"
            style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
            tabindex="-1"
          >
            <div
              class="css-1syq16t"
            >
              <p
                class="chakra-text css-0"
              >
                Mayara Oliveira
              </p>
              <p
                class="chakra-text css-89mcmc"
              >
                Desenvolvedora de Software
              </p>
            </div>
            <hr
              aria-orientation="horizontal"
              class="chakra-menu__divider css-12vdn2q"
            />
            <button
              class="chakra-menu__menuitem css-18esm8n"
              data-index="0"
              id="menu-list-:r2:-menuitem-:r3:"
              role="menuitem"
              tabindex="-1"
              type="button"
            >
              <a
                class="chakra-link css-spn4bz"
                href="https://www.linkedin.com/in/mayoliveii/"
                rel="noopener"
                target="_blank"
              >
                <div
                  class="chakra-stack css-1uodvt1"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"
                    />
                  </svg>
                  <p
                    class="chakra-text css-0"
                  >
                    LinkedIn
                  </p>
                </div>
              </a>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
